### PR TITLE
Introduce implicit last expression results

### PR DESCRIPTION
**Pull request by hamishknight**
Allow implicitly treating the last expression of a function/closure as the result, in addition to the last expression of an if/switch/do expression branch. This is gated behind the experimental feature `ImplicitLastExprResults`.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please build toolchain

**Comment by hamishknight** [comment]
FWIW I still need to add some extra tests

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please build toolchain

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/8051

@swift-ci please build toolchain

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test macOS

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/8051

@swift-ci please build toolchain Windows

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test macOS

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test macOS

**Comment by airspeedswift** [comment]
https://github.com/apple/llvm-project/pull/8051

@swift-ci please build toolchain

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please build toolchain

**Comment by hamishknight** [comment]
@swift-ci please SourceKit stress test

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please build toolchain

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please build toolchain

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please build toolchain

**Comment by hamishknight** [comment]
@swift-ci please test macOS

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test Windows

**Comment by hamishknight** [comment]
Source compat failures are the same as on main

**Review by ahoppen — COMMENTED** [review]
Code completion changes in https://github.com/apple/swift/pull/71006/commits/91c01771ef00b72d920da099409f65a29c492420 look good to me

File: include/swift/Sema/SyntacticElementTarget.h [hunk 1]
```diff
@@ -457,6 +464,11 @@ class SyntacticElementTarget {
     return cast<ExprPattern>(expression.pattern);
   }
 
+  ReturnStmt *getReturnStmt() const {
```
Comment (xedin) [inline include/swift/Sema/SyntacticElementTarget.h:467]: Maybe `getParentReturnStmt()` instead to indicate that we are returning contextual node that wraps the expression?

**Review by xedin — APPROVED** [review]

**Comment by hamishknight** [comment]
@swift-ci please test

**Pull request merged**